src/ir/compilation_unit/path.rs

```diff
@@ -58,6 +58,17 @@ impl TypePath {
         }
     }
 
+    pub fn to_delimited_string(&self, ns_delimiter: String, name_delimiter: String) -> String {
+        let mut string = "".to_string();
+        for element in self.path.0.clone() {
+            string = format!("{}{}{}", string, ns_delimiter, element);
+        }
+        if string.len() > 0 {
+            string = string[1..].to_string();
+        }
+        format!("{}{}{}", string, name_delimiter, self.name)
```

Comment:
All of this does a lot of unnecessary copying (every format! copies all the string data and creates a whole new String).

Using iterators, this can be implemented a lot simpler:

```rust
use ::itertools::Itertools;

let mut string = self.path.0.iter()
    .join(ns_delimiter);

string.push_str(name_delimiter);
string.push_str(&self.name);

string
```

Reply:
Also worth noting, using a delimiter which is allowed in names actually makes this ambiguous. It is possible to craft two separate types that have the same name.

We should probably have the type id in the name, just to avoid that ambiguity.